src/main/java/edu/harvard/iq/dataverse/engine/command/impl/ReconcileDatasetPidCommand.java

```diff
@@ -0,0 +1,181 @@
+        List<RoleAssignment> ras = ctxt.roles().directRoleAssignments(getDataset());
+        for (RoleAssignment ra : ras) {
+            for (AuthenticatedUser au : ctxt.roleAssignees().getExplicitUsers(ctxt.roleAssignees().getRoleAssignee(ra.getAssigneeIdentifier()))) {
+                ctxt.notifications().sendNotification(au, new Timestamp(new Date().getTime()), UserNotification.Type.PIDRECONCILED,getDataset().getLatestVersion().getId(),"Persistent identifier changed!");
+            }
+        }
+
+        return getDataset();
+    }
+
+    private void reconcileFilePids(CommandContext ctxt) {
+        // remove datafile PIDs
+            try {
+                for (DataFile df : getDataset().getFiles()) {
+                    GlobalId oldPid=df.getGlobalId();
+                    PidProvider currentPidProvider = PidUtil.getPidProvider(oldPid.getProviderId());
+                    if (currentPidProvider.alreadyRegistered(df)) {
+                        currentPidProvider.deleteIdentifier(df);// delete it external
+                    }
+
+                    df.setGlobalId(null); // and remove it internally from data structure
+                    df.setGlobalIdCreateTime(null);
+                    df.setIdentifierRegistered(false);
+
+                    AlternativePersistentIdentifier api;
+                    api = new AlternativePersistentIdentifier();
+                    api.setProtocol(oldPid.getProtocol());
+                    api.setAuthority(oldPid.getAuthority());
+                    api.setIdentifier(oldPid.getIdentifier());
+                    api.setDvObject(df);
+                    api.setStorageLocationDesignator(true); // cf. Dataset#getIdentifierForFileStorage()
+                    if(df.getAlternativePersistentIndentifiers()!=null) {
+                        df.getAlternativePersistentIndentifiers().add(api);
+                    }else{
+                        df.setAlternativePersistentIndentifiers(Set.of(api));
+                    }
+                    newPidProvider.generatePid(df); // this updates Protocol, Authority, and Identifier and thus a new GlobalID
+                    logger.fine("Reconciling datafile( id =`" + df.getId() + ")` - creating new globalId `" +df.getGlobalId() + '`');
+                    if (!newPidProvider.registerWhenPublished()) {
+                        registerExternalIdentifier(df, ctxt, true); // this updates GlobalIdCreateTime and IdentifierRegistered
+                    }
+                }
+            } catch (Exception e) {
+                logger.log(Level.WARNING, "Reconcilation of the datafile persistent identifier was not successful:", e.getMessage());
+            }
+    }
+
+    @Override
+    public boolean onSuccess(CommandContext ctxt, Object r) {
+        //update search index with the state
+        ctxt.index().asyncIndexDataset(getDataset(), true);
+        //invalidate all existing Export caches
```

Comment:
I'd suggest removing - we have this code elsewhere for if/when we need it.

Reply:
I removed the comment about export caches. However, the call to reindex is left in place.